File: frontend/lib/data/models/battle_royale_room_model.dart
import 'battle_royale_player_model.dart';

enum RoomStatus { waiting, starting, inProgress, finished }

class BattleRoyaleRoom {
  final String id;
  final String name;
  final String code;
  final String? password;
  final int maxPlayers;
  final int currentPlayers;
  final String hostId;
  final int pairCount;
  final int softCapTime;
  final int? hardCapTime;
  final int flipLimit;
  final String? seed;
  final String region;
  final RoomStatus status;
  final List<BattleRoyalePlayer> players;
  final DateTime createdAt;
  final DateTime? startedAt;

  BattleRoyaleRoom({
    required this.id,
    required this.name,
    required this.code,
    this.password,
    this.maxPlayers = 8,
    this.currentPlayers = 0,
    required this.hostId,
    this.pairCount = 8,
    this.softCapTime = 120,
    this.hardCapTime,
    this.flipLimit = 2,
    this.seed,
    this.region = 'auto',
    this.status = RoomStatus.waiting,
    this.players = const [],
    required this.createdAt,
    this.startedAt,
  });

  factory BattleRoyaleRoom.fromJson(Map<String, dynamic> json) {
    return BattleRoyaleRoom(
      id: json['id'] ?? json['_id'] ?? '',
      name: json['name'] ?? 'Room',
      code: json['code'] ?? '',
      password: json['password'],
      maxPlayers: json['maxPlayers'] ?? 8,
      currentPlayers: json['currentPlayers'] ?? 0,
      hostId: json['hostId'] ?? '',
      pairCount: json['pairCount'] ?? 8,
      softCapTime: json['softCapTime'] ?? 120,
      hardCapTime: json['hardCapTime'],
      flipLimit: json['flipLimit'] ?? 2,
      seed: json['seed'],
      region: json['region'] ?? 'auto',
      status: _parseStatus(json['status']),
      players:
          (json['players'] as List<dynamic>?)
              ?.map((p) => BattleRoyalePlayer.fromJson(p))
              .toList() ??
          [],
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String(),
      ),
      startedAt: json['startedAt'] != null
          ? DateTime.parse(json['startedAt'])
          : null,
    );
  }

  static RoomStatus _parseStatus(dynamic status) {
    if (status == null) return RoomStatus.waiting;
    if (status is String) {
      switch (status.toLowerCase()) {
        case 'waiting':
          return RoomStatus.waiting;
        case 'starting':
          return RoomStatus.starting;
        case 'inprogress':
        case 'in_progress':
          return RoomStatus.inProgress;
        case 'finished':
          return RoomStatus.finished;
        default:
          return RoomStatus.waiting;
      }
    }
    return RoomStatus.waiting;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'code': code,
      'password': password,
      'maxPlayers': maxPlayers,
      'currentPlayers': currentPlayers,
      'hostId': hostId,
      'pairCount': pairCount,
      'softCapTime': softCapTime,
      'hardCapTime': hardCapTime,
      'flipLimit': flipLimit,
      'seed': seed,
      'region': region,
      'status': status.name,
      'players': players.map((p) => p.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'startedAt': startedAt?.toIso8601String(),
    };
  }

  bool get isFull => currentPlayers >= maxPlayers;
  bool get hasPassword => password != null && password!.isNotEmpty;
  bool get canStart =>
      currentPlayers >= 2 &&
      players.where((p) => p.isReady || p.isHost).length >= 2;

  BattleRoyaleRoom copyWith({
    String? id,
    String? name,
    String? code,
    String? password,
    int? maxPlayers,
    int? currentPlayers,
    String? hostId,
    int? pairCount,
    int? softCapTime,
    int? hardCapTime,
    int? flipLimit,
    String? seed,
    String? region,
    RoomStatus? status,
    List<BattleRoyalePlayer>? players,
    DateTime? createdAt,
    DateTime? startedAt,
  }) {
    return BattleRoyaleRoom(
      id: id ?? this.id,
      name: name ?? this.name,
      code: code ?? this.code,
      password: password ?? this.password,
      maxPlayers: maxPlayers ?? this.maxPlayers,
      currentPlayers: currentPlayers ?? this.currentPlayers,
      hostId: hostId ?? this.hostId,
      pairCount: pairCount ?? this.pairCount,
      softCapTime: softCapTime ?? this.softCapTime,
      hardCapTime: hardCapTime ?? this.hardCapTime,
      flipLimit: flipLimit ?? this.flipLimit,
      seed: seed ?? this.seed,
      region: region ?? this.region,
      status: status ?? this.status,
      players: players ?? this.players,
      createdAt: createdAt ?? this.createdAt,
      startedAt: startedAt ?? this.startedAt,
    );
  }
}

File: backend/src/models/battleRoyaleRoomModel.js
const mongoose = require("mongoose");

const playerSchema = new mongoose.Schema({
	userId: {type: mongoose.Schema.Types.ObjectId, ref: "User", required: true},
	username: {type: String, required: true},
	avatarUrl: {type: String, default: null},
	borderColor: {type: String, default: "#4CAF50"},
	isReady: {type: Boolean, default: false},
	isHost: {type: Boolean, default: false},
	ping: {type: Number, default: null},
	isConnected: {type: Boolean, default: true},
	disconnectedAt: {type: Date, default: null},
	socketId: {type: String, default: null},
	joinedAt: {type: Date, default: Date.now},
});

const battleRoyaleRoomSchema = new mongoose.Schema({
	code: {type: String, required: true, unique: true, index: true},
	name: {type: String, required: true},
	password: {type: String, default: null},
	maxPlayers: {type: Number, default: 8, min: 2, max: 8},
	currentPlayers: {type: Number, default: 0},
	hostId: {type: mongoose.Schema.Types.ObjectId, ref: "User", required: true},
	pairCount: {type: Number, default: 8},
	softCapTime: {type: Number, default: 120},
	hardCapTime: {type: Number, default: null},
	flipLimit: {type: Number, default: 2},
	seed: {type: String, default: null},
	region: {type: String, default: "auto"},
	status: {
		type: String,
		enum: ["waiting", "starting", "inProgress", "finished"],
		default: "waiting",
	},
	players: [playerSchema],
	matchId: {
		type: mongoose.Schema.Types.ObjectId,
		ref: "BattleRoyaleMatch",
		default: null,
	},
	createdAt: {type: Date, default: Date.now},
	startedAt: {type: Date, default: null},
	finishedAt: {type: Date, default: null},
	updatedAt: {type: Date, default: Date.now},
});

battleRoyaleRoomSchema.index({status: 1});
battleRoyaleRoomSchema.index({hostId: 1});
battleRoyaleRoomSchema.index({createdAt: -1});
battleRoyaleRoomSchema.index({"players.userId": 1});

battleRoyaleRoomSchema.pre("save", function (next) {
	this.currentPlayers = this.players.filter((p) => p.isConnected).length;
	this.updatedAt = Date.now();
	next();
});

battleRoyaleRoomSchema.methods.isFull = function () {
	return this.currentPlayers >= this.maxPlayers;
};

battleRoyaleRoomSchema.methods.canStart = function () {
	const nonHostPlayers = this.players.filter(
		(p) => !p.isHost && p.isConnected
	);
	const allNonHostReady = nonHostPlayers.every((p) => p.isReady);
	return this.currentPlayers >= 2 && allNonHostReady;
};

battleRoyaleRoomSchema.methods.getPlayer = function (userId) {
	return this.players.find((p) => p.userId.toString() === userId.toString());
};

battleRoyaleRoomSchema.methods.addPlayer = function (
	userId,
	username,
	socketId,
	avatarUrl = null,
	borderColor = "#4CAF50"
) {
	const existingPlayer = this.getPlayer(userId);
	if (existingPlayer) {
		existingPlayer.isConnected = true;
		existingPlayer.socketId = socketId;
		existingPlayer.disconnectedAt = null;
		if (avatarUrl) existingPlayer.avatarUrl = avatarUrl;
		if (borderColor) existingPlayer.borderColor = borderColor;
	} else {
		this.players.push({
			userId,
			username,
			socketId,
			avatarUrl,
			borderColor,
			isHost: this.players.length === 0,
			isConnected: true,
		});
	}
};

battleRoyaleRoomSchema.methods.removePlayer = function (userId) {
	const playerIndex = this.players.findIndex(
		(p) => p.userId.toString() === userId.toString()
	);
	if (playerIndex !== -1) {
		const wasHost = this.players[playerIndex].isHost;
		this.players.splice(playerIndex, 1);

		if (wasHost && this.players.length > 0) {
			const newHost = this.players
				.filter((p) => p.isConnected)
				.sort((a, b) => a.joinedAt - b.joinedAt)[0];
			if (newHost) {
				newHost.isHost = true;
				this.hostId = newHost.userId;
			}
		}
	}
};

module.exports = mongoose.model("BattleRoyaleRoom", battleRoyaleRoomSchema);
